ui/pages/ApplicationAdd.tsx

```diff
@@ -231,22 +246,36 @@ function AddApplication({ className }: Props) {
             </FormHelperText>
           </FormElement>
           <FormElement>
-            <TextField
+            <RepoInputWithAuth
+              isAuthenticated={!!formState.url && credentialsDetected}
               onChange={(e) => {
                 setFormState({
                   ...formState,
                   url: e.currentTarget.value,
                 });
               }}
+              onProviderChange={(provider: GitProvider) => {
+                setFormState({ ...formState, provider });
+              }}
+              onAuthClick={(provider) => {
+                if (provider === GitProvider.GitHub) {
+                  setAuthOpen(true);
+                }
+
+                if (provider === GitProvider.GitLab) {
+                  storeCallbackState({
+                    page: PageRoute.ApplicationAdd,
+                    state: formState,
+                  });
+                }
+              }}
               required
               id="url"
               label="Source Repo URL"
               variant="standard"
               value={formState.url}
+              helperText="The git repository URL where the application YAML files are stored"
```

Comment:
Are we going to have to put all the new auth stuff in this component into every component where we do some sort of action? It makes me want to put all this auth stuff at the highest level right? Will we know what provider the user is with as a part of the application object?

Reply:
At the moment, the GitHub auth stuff isn't encapsulated anywhere. The GitLab button is pretty well encapsulated, as anywhere you drop it in It Should Work:tm:.

I would like to get to the point where anywhere you drop in the`GitHubAuthButton` it opens the modal and does the whole thing. This would have required a larger change, however, and made this PR tougher to review.

>  Will we know what provider the user is with as a part of the application object?

We can check the URL(s) of the `Application` to figure out which provider it is.